helix-core/src/main/java/org/apache/helix/model/TrieClusterTopology.java

```diff
@@ -0,0 +1,244 @@
+          nodeStack.push(child);
+        }
+      }
+    }
+    return resultMap;
+  }
+
+  private TrieNode getStartNode(LinkedHashMap<String, String> domain) {
+    TrieNode curNode = _rootNode;
+    TrieNode nextNode;
+    for (Map.Entry<String, String> entry : domain.entrySet()) {
+      nextNode = curNode.getChildren().get(entry.getKey() + CONNECTOR + entry.getValue());
+      if (nextNode == null) {
+        throw new IllegalArgumentException(String
+            .format("The input domain %s does not have the value %s", entry.getKey(),
+                entry.getValue()));
+      }
+      curNode = nextNode;
+    }
+    return curNode;
+  }
+
+  private void removeInvalidInstanceConfig(Map<String, InstanceConfig> instanceConfigMap) {
+    Set<String> toRemoveConfig = new HashSet<>();
+    for (String instanceName : instanceConfigMap.keySet()) {
+      Map<String, String> domainAsMap = instanceConfigMap.get(instanceName).getDomainAsMap();
+      if (domainAsMap.isEmpty()) {
+        logger.info(String.format("Domain for instance %s is not set", instanceName));
+        toRemoveConfig.add(instanceName);
+      } else {
+        for (String key : _topologyKeys) {
+          String value = domainAsMap.get(key);
+          if (value == null || value.length() == 0) {
+            logger.info(String.format("Domain %s for instance %s is not set",
+                domainAsMap.get(key), instanceName));
+            toRemoveConfig.add(instanceName);
+            break;
+          }
+        }
+      }
+    }
+    toRemoveConfig.forEach(entry -> instanceConfigMap.remove(entry));
+  }
+
+  /**
+   * Constructs a trie based on the provided instance config map. It loops through all instance
+   * configs and constructs the trie in a top down manner.
+   */
+  private void constructTrie(Map<String, InstanceConfig> instanceConfigMap) {
+    removeInvalidInstanceConfig(instanceConfigMap);
+    Map<String, Map<String, String>> instanceDomainsMap = new HashMap<>();
+    instanceConfigMap.entrySet().forEach(
+        entry -> instanceDomainsMap.put(entry.getKey(), entry.getValue().getDomainAsMap()));
+
+    for (Map.Entry<String, Map<String, String>> entry : instanceDomainsMap.entrySet()) {
+      TrieNode curNode = _rootNode;
+      String path = "";
+      for (int i = 0; i < _topologyKeys.length; i++) {
+        String key = _topologyKeys[i] + CONNECTOR + entry.getValue().get(_topologyKeys[i]);
+        path = path + DELIMITER + key;
+        TrieNode nextNode = curNode.getChildren().get(key);
+        if (nextNode == null) {
+          nextNode = new TrieNode(new HashMap<>(), path);
+        }
+        curNode.addChild(key, nextNode);
+        curNode = nextNode;
+      }
+    }
+  }
+
+  private static class TrieNode {
+    // A mapping between trie key and children nodes.
+    private Map<String, TrieNode> _children;
+
+    // the complete path/prefix leading to the current node.
+    private final String _path;
```

Comment:
This API is for the common case where instances are returned for the fault zone type:
  public Map<String, List<String>> getInstancesUnderFaultZone() {
    return getTopologyUnderDomainType(_faultZoneType);
  }
The name of the instance is implicitly embedded in the `path` field of the Trie node. e.g. the path could be like zone:myzone0/instance:myinstance0, so 'myinstance0' is the instance name.
I didn't see the necessity of marking leaf nodes. Do you think it'll save time in querying? Because we always need to traverse to the leaves anyway.